### ISSUE TITLE
Driving skill plugin avoidCollision error for linked roads with a different number of lanes

### ISSUE THREAD

**Issue by roncolic** [gama-platform/gama#3112]
**Describe the bug**
I'm using the Driving Skill plugin and i extend it. To achieve this I need new shapefiles with detailed informations about the lanes and their direction. Some of my roads have fewer lanes than their linked road. When I launch experiment_city, after a few cycles, I get this error : 
https://imgur.com/a/WDp1dDL

**To Reproduce**
Use the Driving Skill plugin model.
I recommend to create more bidirectional roads with different numbers of lanes with the following since there aren't any by default.

```
		create road from: shape_file_roads with: [lanes::int(read("lanes")), oneway::string(read("oneway"))] {
			geom_display <- shape + (2.5 * lanes);
			maxspeed <- (lanes = 1 ? 30.0 : (lanes = 2 ? 50.0 : 70.0)) °km / °h;
			switch oneway {
				match_one ["no","",nil] {
					create road {
						lanes <- 2;//max([1, int(myself.lanes / 2.0)]);
						shape <- polyline(reverse(myself.shape.points));
						maxspeed <- myself.maxspeed;
						geom_display <- myself.geom_display;
						linked_road <- myself;
						myself.linked_road <- self;
					}

					lanes <- 1;//int(lanes / 2.0 + 0.5);
				}

				match "-1" {
					shape <- polyline(reverse(shape.points));
				}

			}

		}
```

**Desktop (please complete the following information):**
 - OS: Windows 10
 - GAMA version: GAMA_1.8.1_Windows_with_JDK

**Comment :**
This error doesn't appear when the number of lanes on both sides of the roads are equal.
The line  `final IList aglanes = (IList) ((IList) currentRoad.getAttribute(RoadSkill.AGENTS_ON)).get(lane);` is responsible for the error. I’m not sure what it’s supposed to do, but the value `nbLinkedLanes - 1` is affected to the parameter `lane` at line 1587 while `linkedRoad` isn’t affected to `currentRoad` anywhere. I suppose the highest index of the linkedRoad lanes being used with the currentRoad (which has fewer lanes) is the cause.

Thank you in advance.

**Comment by minhduc0711** [comment]
Hello, some new updates have landed in GAMA 1.8.2, which includes some fixes related to linked roads. Old models should work with the new version.
Can you try your model on the alpha release?

**Comment by roncolic** [comment]
Hi, thank you for your response.
I tried my code with GAMA 1.8.2 and got the following error immediatly after trying to run the experiment :
https://imgur.com/a/bmUttnJ
I tried `drive_random` instead of `drive` but the result was the same.

**Comment by minhduc0711** [comment]
I believe this is caused by this line in your model
```
create road from: shape_file_roads with: [lanes::int(read("lanes")), oneway::string(read("oneway"))] {
```
Can you try replacing `lanes::int(read("lanes"))` with `num_lanes::int(read("lanes"))`?

**Comment by roncolic** [comment]
Seems like that fixed it.
However, now I get : https://imgur.com/a/cFVqse3
I got the first error the first time i ran the experiment but i only get the second error since then.

**Comment by minhduc0711** [comment]
Can you share the code for your model?

**Comment by roncolic** [comment]
Forget about my previous comment, I made some changes and solved (?) this issue. I believed it was caused by the skill `skill_road_node` being inherited (since the error doesn't show up when my species declare it direclty). Are skills not supposed to be inherited ?

Another error appeared (thanks for telling me i can upload images) :
![image](https://user-images.githubusercontent.com/40756603/124615585-b4c0b480-de75-11eb-9e83-1e5be352abe7.png)
Not sure if I can share the code for my model but the same problem happens with the old Driving skill model, just need to replace `lanes` with `num_lanes` : 
https://github.com/gama-platform/gama/blob/master/simtools.gaml.extensions.traffic/models/Driving%20Skill/models/Driving%20skill.gaml

**Comment by minhduc0711** [comment]
> Are skills not supposed to be inherited ?
You meant you subclasssed `RoadNodeSkill` in Java? I think there should be no problem. Would be great if you can share a minimal code snippet just for reproducing the issue.

> the same problem happens with the old Driving skill model
I'm using `GAMA_1.8.2_Alpha_Linux_07.05.21_86ad2069.zip` for the old model but I can't seem to reproduce the error. Are the new example models broken for you too?

**Comment by roncolic** [comment]
I had a problem with this :
```
species road_node skills: [skill_road_node] {...}

species intersection parent: road_node {...}

species bus_stop parent: road_node {...}
```
And apparently switching to this fixed it :
```
species intersection skills: [skill_road_node]{...}

species bus_stop skills: [skill_road_node]{...}
```

I'm using `GAMA_1.8.2_Alpha_withJDK_Win_06.29.21_2b9568ed.zip`. I only tried the ring experiment from the Drive Random model in the new Driving Skill and it works.

**Comment by roncolic** [comment]
After restarting gama this is what I get : 
![image](https://user-images.githubusercontent.com/40756603/124731451-880eaa80-df12-11eb-83e9-ce9dc780cb29.png)

**Comment by minhduc0711** [comment]
Can you try the newer build `86ad2069`?

**Comment by roncolic** [comment]
Tried with the newer build, got : 
![image](https://user-images.githubusercontent.com/40756603/124744437-0a9d6700-df1f-11eb-894e-89da74b4424b.png)

**Comment by minhduc0711** [comment]
I issued some fixes for the problems you reported. You can try again once the new build is released.

**Comment by minhduc0711** [comment]
Can you try again on the new release 1dfe2b6?

**Comment by roncolic** [comment]
Just tried it, i don't get any errors anymore.
However I noticed on the 1.8.1 version that after some time (a few thousand cycles) the vehicles unexpectedly stopped on the roads without any error being raised. The only things I noticed are differences between the road attributes `all_agents` and `agents_on`.

![problème_cas_1_cropped](https://user-images.githubusercontent.com/40756603/126648749-f759519b-434c-400f-8c86-00725831b09a.png)

![problème_cas_2_cropped](https://user-images.githubusercontent.com/40756603/126648776-c52a32b1-4a0c-446f-bc80-3bbf56566129.png)

I was wondering if the linked roads were not responsible since the vehicles wrongly registered in `agents_on` sometimes never even used the road but always used the linked road (in my tests at least).

 With the latest release it seems to happen even more but I am not sure if it's the same problem.

Should I open a new issue for this ?
